src/webserver/routes/api/reports.ts

```diff
@@ -0,0 +1,645 @@
+						Validators.and(Validators.integer, Validators.gte(1)),
+					),
+					skip: Validators.optional(
+						Validators.and(Validators.integer, Validators.gte(1)),
+					),
+					csv: Validators.optional(Validators.booleanString),
+				},
+			},
+		}),
+	async (req, res) => {
+		let from = dateFromData(req.query.from);
+		let to = dateFromData(req.query.to);
+		let count = numberFromData(req.query.count) || undefined;
+		let skip = numberFromData(req.query.skip) || 0;
+
+		let query: mongoose.FilterQuery<ITransaction> = {};
+		if (from || to) query.date = {};
+		if (from) query.date.$gte = from;
+		if (to) query.date.$lte = to;
+
+		let transactions = await Transaction.find(query, null, {
+			limit: count,
+			skip: skip,
+		});
+
+		let data = await Promise.all(transactions.map(x => x.toAPIResponse()));
+
+		if (req.query.csv && booleanFromData(req.query.csv)) {
+			let csv = await json2csv.json2csvAsync(data, {
+				keys: [
+					{
+						field: '_id',
+						title: 'ID',
+					},
+					{
+						field: 'date',
+						title: 'Date',
+					},
+					{
+						field: 'from.text',
+						title: 'From Name',
+					},
+					{
+						field: 'from.id',
+						title: 'From ID',
+					},
+					{
+						field: 'to.text',
+						title: 'To Name',
+					},
+					{
+						field: 'to.id',
+						title: 'To ID',
+					},
+					{
+						field: 'amount',
+						title: 'Amount',
+					},
+					{
+						field: 'reason',
+						title: 'Reason',
+					},
+				],
+				emptyFieldValue: '',
+			});
+			res.setHeader('Content-Type', 'text/csv');
+			res.setHeader(
+				'Content-Disposition',
+				'attachment; filename="alltransactions.csv"',
+			);
+			res.send(csv);
+			return;
+		}
+
+		res.status(200).json(data);
+	},
+);
+
+router.get(
+	'/balance/total',
+	async (req, res, next) =>
+		request(req, res, next, {
+			authentication: true,
+			roles: ['ADMIN'],
+		}),
+	async (req, res) => {
+		let [{balance}] = await User.aggregate([
+			{
+				$match: {
+					$and: [
+						{
+							roles: {$bitsAllSet: UserRoles.STUDENT},
+						},
+						{
+							roles: {$bitsAllClear: UserRoles.STAFF},
+						},
+					],
+				},
+			},
+			{
+				$group: {
+					_id: null,
+					balance: {$sum: '$balance'},
+				},
+			},
+		]);
+
+		balance = roundCurrency(balance);
+
+		res.status(200).json({balance});
+	},
+);
+
+router.get(
+	'/balance/top',
+	async (req, res, next) =>
+		request(req, res, next, {
+			authentication: true,
+			roles: ['ADMIN'],
+			validators: {
+				query: {
+					count: Validators.optional(
+						Validators.and(Validators.integer, Validators.gte(1)),
+					),
+					csv: Validators.optional(Validators.booleanString),
+				},
+			},
+		}),
+	async (req, res) => {
+		let count = numberFromData(req.query.count) || 10;
+		let topUsersQuery = User.find({
+			$and: [
+				{
+					roles: {$bitsAllSet: UserRoles.STUDENT},
+				},
+				{
+					roles: {$bitsAllClear: UserRoles.STAFF},
+				},
+			],
+		}).sort({balance: -1});
+
+		let topUsers =
```

Comment:
This should have a filter so that it only returns users with non-zero balances.

Reply:
I'll work on getting this changed. Should users with a balance of 0 still show up in the spreadsheet?

Reply:
> Well is the spreadsheet going to list every user or just the top x users?

The current behavior is that the spreadsheet lists all users sorted by balance, but it could be changed to only list users with a balance greater than 1

Reply:
I think users with zero balance shouldn't be on it. Otherwise it's an export of all users sorted by balance.